// gatsby-remark-embed-gist/gatsby-ssr.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.onRenderBody = onRenderBody;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

/**
 * @typedef {object} PluginOptions
 * @property {string} username the default gist user.
 * @property {boolean} includeDefaultCss a flag indicating the default css should be included
 */

/**
 * Adds a link to the Github Gist default css.
 * @param {{ setHeadComponents }} setHeadComponents adds components to <head>.
 * @param {PluginOptions} options the options of the plugin.
 * @returns {*} rendered body.
 */
function onRenderBody(_ref) {
  var setHeadComponents = _ref.setHeadComponents;
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  options = _objectSpread(_objectSpread({}, {
    gistCssPreload: false,
    gistCssUrlAddress: "https://github.githubassets.com/assets/gist-embed-b3b573358bfc66d89e1e95dbf8319c09.css"
  }), options);
  var includeCss = true;

  if (options.gistDefaultCssInclude != null) {
    includeCss = options.gistDefaultCssInclude;
  } else if (options.includeDefaultCss != null) {
    includeCss = options.includeDefaultCss;
  }

  var key = "gist-embed-b3b573358bfc66d89e1e95dbf8319c09";

  if (includeCss) {
    setHeadComponents(options.gistCssPreload ? [/*#__PURE__*/_react["default"].createElement("link", {
      id: key,
      as: "style",
      href: options.gistCssUrlAddress,
      key: key,
      rel: "preload"
    }), /*#__PURE__*/_react["default"].createElement("noscript", {
      key: "noscript-" + key
    }, /*#__PURE__*/_react["default"].createElement("link", {
      href: options.gistCssUrlAddress,
      rel: "stylesheet"
    })), /*#__PURE__*/_react["default"].createElement("script", {
      dangerouslySetInnerHTML: {
        __html: "\n                  var link = document.querySelector(\"#".concat(key, "\");\n                  link.addEventListener(\"load\", function() {\n                    this.rel = \"stylesheet\"\n                  });\n                ")
      }
    })] : [/*#__PURE__*/_react["default"].createElement("link", {
      href: options.gistCssUrlAddress,
      key: key,
      rel: "stylesheet"
    })]);
  }
}
