// @contentful/rich-text-react-renderer/dist/lib/__test__/documents/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var hr_1 = require("./hr");
exports.hrDoc = hr_1.default;
var hyperlink_1 = require("./hyperlink");
exports.hyperlinkDoc = hyperlink_1.default;
var invalid_marks_1 = require("./invalid-marks");
exports.invalidMarksDoc = invalid_marks_1.default;
var invalid_type_1 = require("./invalid-type");
exports.invalidTypeDoc = invalid_type_1.default;
var paragraph_1 = require("./paragraph");
exports.paragraphDoc = paragraph_1.default;
var heading_1 = require("./heading");
exports.headingDoc = heading_1.default;
var mark_1 = require("./mark");
exports.marksDoc = mark_1.default;
var multi_mark_1 = require("./multi-mark");
exports.multiMarkDoc = multi_mark_1.default;
var embedded_entry_1 = require("./embedded-entry");
exports.embeddedEntryDoc = embedded_entry_1.default;
var inline_entity_1 = require("./inline-entity");
exports.inlineEntityDoc = inline_entity_1.default;
var ol_1 = require("./ol");
exports.olDoc = ol_1.default;
var ul_1 = require("./ul");
exports.ulDoc = ul_1.default;
var quote_1 = require("./quote");
exports.quoteDoc = quote_1.default;
