// react-instantsearch-core/dist/cjs/connectors/connectDynamicWidgets.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _createConnector = _interopRequireDefault(require("../core/createConnector"));

var _indexUtils = require("../core/indexUtils");

var _default = (0, _createConnector.default)({
  displayName: 'AlgoliaDynamicWidgets',
  defaultProps: {
    transformItems: function transformItems(items) {
      return items;
    }
  },
  propTypes: {
    transformItems: _propTypes.default.func
  },
  getProvidedProps: function getProvidedProps(props, _searchState, searchResults) {
    var results = (0, _indexUtils.getResults)(searchResults, {
      ais: props.contextValue,
      multiIndexContext: props.indexContextValue
    });

    if (!results) {
      return {
        attributesToRender: []
      };
    }

    var facetOrder = results.renderingContent && results.renderingContent.facetOrdering && results.renderingContent.facetOrdering.facets && results.renderingContent.facetOrdering.facets.order || [];
    return {
      attributesToRender: props.transformItems(facetOrder, {
        results: results
      })
    };
  }
});

exports.default = _default;
