// react-instantsearch/dist/dom.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "InstantSearch", {
  enumerable: true,
  get: function get() {
    return _reactInstantsearchDom.InstantSearch;
  }
});
Object.defineProperty(exports, "Index", {
  enumerable: true,
  get: function get() {
    return _reactInstantsearchDom.Index;
  }
});
Object.defineProperty(exports, "Breadcrumb", {
  enumerable: true,
  get: function get() {
    return _reactInstantsearchDom.Breadcrumb;
  }
});
Object.defineProperty(exports, "ClearRefinements", {
  enumerable: true,
  get: function get() {
    return _reactInstantsearchDom.ClearRefinements;
  }
});
Object.defineProperty(exports, "Configure", {
  enumerable: true,
  get: function get() {
    return _reactInstantsearchDom.Configure;
  }
});
Object.defineProperty(exports, "ExperimentalConfigureRelatedItems", {
  enumerable: true,
  get: function get() {
    return _reactInstantsearchDom.ExperimentalConfigureRelatedItems;
  }
});
Object.defineProperty(exports, "ExperimentalDynamicWidgets", {
  enumerable: true,
  get: function get() {
    return _reactInstantsearchDom.ExperimentalDynamicWidgets;
  }
});
Object.defineProperty(exports, "CurrentRefinements", {
  enumerable: true,
  get: function get() {
    return _reactInstantsearchDom.CurrentRefinements;
  }
});
Object.defineProperty(exports, "HierarchicalMenu", {
  enumerable: true,
  get: function get() {
    return _reactInstantsearchDom.HierarchicalMenu;
  }
});
Object.defineProperty(exports, "Highlight", {
  enumerable: true,
  get: function get() {
    return _reactInstantsearchDom.Highlight;
  }
});
Object.defineProperty(exports, "Hits", {
  enumerable: true,
  get: function get() {
    return _reactInstantsearchDom.Hits;
  }
});
Object.defineProperty(exports, "HitsPerPage", {
  enumerable: true,
  get: function get() {
    return _reactInstantsearchDom.HitsPerPage;
  }
});
Object.defineProperty(exports, "InfiniteHits", {
  enumerable: true,
  get: function get() {
    return _reactInstantsearchDom.InfiniteHits;
  }
});
Object.defineProperty(exports, "Menu", {
  enumerable: true,
  get: function get() {
    return _reactInstantsearchDom.Menu;
  }
});
Object.defineProperty(exports, "MenuSelect", {
  enumerable: true,
  get: function get() {
    return _reactInstantsearchDom.MenuSelect;
  }
});
Object.defineProperty(exports, "NumericMenu", {
  enumerable: true,
  get: function get() {
    return _reactInstantsearchDom.NumericMenu;
  }
});
Object.defineProperty(exports, "Pagination", {
  enumerable: true,
  get: function get() {
    return _reactInstantsearchDom.Pagination;
  }
});
Object.defineProperty(exports, "Panel", {
  enumerable: true,
  get: function get() {
    return _reactInstantsearchDom.Panel;
  }
});
Object.defineProperty(exports, "PoweredBy", {
  enumerable: true,
  get: function get() {
    return _reactInstantsearchDom.PoweredBy;
  }
});
Object.defineProperty(exports, "RangeInput", {
  enumerable: true,
  get: function get() {
    return _reactInstantsearchDom.RangeInput;
  }
});
Object.defineProperty(exports, "RangeSlider", {
  enumerable: true,
  get: function get() {
    return _reactInstantsearchDom.RangeSlider;
  }
});
Object.defineProperty(exports, "RatingMenu", {
  enumerable: true,
  get: function get() {
    return _reactInstantsearchDom.RatingMenu;
  }
});
Object.defineProperty(exports, "RefinementList", {
  enumerable: true,
  get: function get() {
    return _reactInstantsearchDom.RefinementList;
  }
});
Object.defineProperty(exports, "ScrollTo", {
  enumerable: true,
  get: function get() {
    return _reactInstantsearchDom.ScrollTo;
  }
});
Object.defineProperty(exports, "SearchBox", {
  enumerable: true,
  get: function get() {
    return _reactInstantsearchDom.SearchBox;
  }
});
Object.defineProperty(exports, "RelevantSort", {
  enumerable: true,
  get: function get() {
    return _reactInstantsearchDom.RelevantSort;
  }
});
Object.defineProperty(exports, "Snippet", {
  enumerable: true,
  get: function get() {
    return _reactInstantsearchDom.Snippet;
  }
});
Object.defineProperty(exports, "SortBy", {
  enumerable: true,
  get: function get() {
    return _reactInstantsearchDom.SortBy;
  }
});
Object.defineProperty(exports, "Stats", {
  enumerable: true,
  get: function get() {
    return _reactInstantsearchDom.Stats;
  }
});
Object.defineProperty(exports, "ToggleRefinement", {
  enumerable: true,
  get: function get() {
    return _reactInstantsearchDom.ToggleRefinement;
  }
});
Object.defineProperty(exports, "getInsightsAnonymousUserToken", {
  enumerable: true,
  get: function get() {
    return _reactInstantsearchDom.getInsightsAnonymousUserToken;
  }
});

var _reactInstantsearchDom = require("react-instantsearch-dom");
